### PR TITLE
Cache folded syntax tree

### PR DESCRIPTION
**Pull request by marcelofabri**

### CODE REVIEW

**Comment by marcelofabri** [comment]
I was hoping to get more speed gains tbh, not sure if it's worth merging this. Probably it is given that the number of rules using `folded()` will in theory only go up?

**Review by SimplyDanny — APPROVED** [review]

**Comment by SimplyDanny** [comment]
Makes sense to me. I expect more rules using it over time.

**Comment by marcelofabri** [comment]
Merging because this also has the nice benefit of allowing a custom rule to skip validation based on the file path by returning `nil` in `preprocess`.

File: Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift [hunk 1]
```diff
@@ -24,7 +24,7 @@ public protocol SwiftSyntaxRule: SourceKitFreeRule {
     /// - parameter syntaxTree: The syntax tree to run pre-processing on
```
Comment (jpsim) [inline Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift:24]: This is now out of date.

**Review by jpsim — COMMENTED** [review]
+1 other than the outdated docs

File: Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift (hunk 1 above)
Comment (marcelofabri) [reply Source/SwiftLintFramework/Protocols/SwiftSyntaxRule.swift:24]: good catch, will update

**Pull request merged**